dds_web/api/project.py

```diff
@@ -188,6 +188,115 @@ def post(self):
 
         return {"message": return_message}
 
+    @auth.login_required(role=["Unit Admin", "Unit Personnel"])
+    @logging_bind_request
+    @json_required
+    @handle_validation_errors
+    def patch(self):
+        """Partially update a the project status"""
+        # Get project ID, project and verify access
+        project_id = dds_web.utils.get_required_item(obj=flask.request.args, req="project")
+        project = dds_web.utils.collect_project(project_id=project_id)
+        dds_web.utils.verify_project_access(project=project)
+
+        # Get json input from request
+        json_input = flask.request.get_json(silent=True)  # Already checked by json_required
+
+        # Operation must be confirmed by the user - False by default
+        confirmed_operation = json_input.get("confirmed", False)
+        if not isinstance(confirmed_operation, bool):
+            raise DDSArgumentError(message="`confirmed` is a boolean value: True or False.")
+        if not confirmed_operation:
+            warning_message = "Operation must be confirmed before proceding."
+            project_info = ProjectInfo().get()
+            project_status = self.get()
+            json_returned = {
+                **project_info,
+                "project_status": project_status,
+                "warning": warning_message,
+                "default_unit_days": project.responsible_unit.days_in_expired,
+            }
+            return json_returned
+
+        # Cannot change project status if project is busy
+        if project.busy:
+            raise ProjectBusyError(
+                message=(
+                    f"The deadline for the project '{project_id}' is already in the process of being changed. "
+                    "Please try again later. \n\nIf you know the project is not busy, contact support."
+                )
+            )
+
+        self.set_busy(project=project, busy=True)
```

Comment:
So since `set_busy` doesn't contain a commit we'll need to add a try except to this and the end where busy is set to False. This is not done in the other `ProjectStatus` endpoints either (noticed now), but if we need to change those too then that's another task. Unless we could add a test (or test manually) somehow to verify that that the `busy=True` is actually noticed in other database sessions running in parallel

Reply:
You mean inserting the set busy inside the try catch below?

Like

```
try:
            self.set_busy(project=project, busy=True)
            logic_of_the_code()
except:
            self.set_busy(project=project, busy=False)
            raise
return {"message": return_message}
```

Reply:
No we need to make sure that the `self.set_busy` is commited first and also that the last change with `set_busy` is commited.
```python
try:
            self.set_busy(project=project, busy=True)
            db.session.commit()
            logic_of_the_code()
except:
            try: 
                 self.set_busy(project=project, busy=False)
            except SQLAlchemyError, OperationalError as err: 
                 # log something 
                 raise
            raise
return {"message": return_message}
```

But also we don't want SQLAlchemyError etc to be raised, we want `DatabaseError` to be raised, e.g. via the decorator that I mentioned in another comment (haven't checked yet if that's changed now)

Reply:
@rv0lt forget this

Reply:
The busy status is actually commited, I forgot. You don't need to change it at all.

Reply:
In fact move the `self.set_busy(project=project, busy=True)` to where you had it before, outside the try except

Reply:
Okay, yes I was getting confused with the _set busy_. Bc with the tests I discovered that mocking one commit actually mocked the set busy. Good to confirm that it actually commits

Reply:
Yeah it's the decorator on the function that does it, I was confusing myself too